### ISSUE TITLE
Support for Codeium autocomplete as a Github Copilot alternative

### ISSUE THREAD

**Issue by justjhong** [rstudio/rstudio#13278]
Hello RStudio team! I recently saw the first-party support for Github Copilot in RStudio (https://github.com/rstudio/rstudio/pull/13192). I'm currently in the process of developing an add-in to support code suggestions from Codeium (https://codeium.com/) a Copilot competitor. However, due to the limitations of the rstudioapi, it would not be possible to have inline ghost text suggestions without modifying the RStudio code itself as shown in the linked PR. The closest alternative for an add-in seems to be a panelViewer containing a preview of completions, which is not ideal.

I was wondering if the RStudio maintainers would be open to supporting Codeium in the same way as Copilot? If there is a way to piggyback on the Copilot PR by using the same ghost text and suggestion functionality, and the team is ok with it, I would be open to trying to make this PR as a contribution.

Thank you for maintaining and improving upon this great open-source platform for all R users!

**Comment by alanpaulkwan** [comment]
Wonderful idea - I love codeium and (1) it is free for non-commercial use unlike Copilot, and (2) pretty darn good with R.

**Comment by kevinushey** [comment]
I think we could consider something like this. We could consider something in `rstudioapi` like `registerInlineCompletionHandler()`, which would take an R function and return something like an insertion range + ghost text to be displayed.

**Comment by justjhong** [comment]
Thanks for the reply @kevinushey! Support for inline ghost text completion suggestions would be ideal for us and avoid having to mess with core studio code. How do you imagine `registerInlineCompletionHandler()` to work? I can think of two options at the moment:

1. The handler takes in document contents and cursor position and is fired on document change. The handler would output an insertion range(s) (note this is plural for inline interpolated completions) and ghost text to be displayed as you mentioned. An open question here is who would handle the tab completion? Ideally we could piggy back on the copilot tab completion implementation, but we would also ideally be able to register a callback on accepting completions as well.
2. Rstudioapi simply exposes functions to add ghost text. This would take in insertion range(s) and ghost text to be displayed. Clearing the completion would be handled the same way as it is for copilot (clicking away, moving the text cursor).

I would love to hear your thoughts on this, and thank you again for considering this issue!

**Comment by kevinushey** [comment]
I think option (1) is probably the better solution, especially if we want to allow Codeium to run more immediately in response to user input, so that the user doesn't necessarily have to wait as long for a completion result. However, it may take more work to hook this up since (at least currently) we don't have plumbing for notifying clients of document diffs; we just have the whole document contents at once.

Option (2) is definitely more straight-forward, but might make tools like Codeium a little bit slower (since they need to build their context and provide a completion result all at once). If that were fast enough in practice, perhaps that would be a better solution?

> An open question here is who would handle the tab completion? Ideally we could piggy back on the copilot tab completion implementation, but we would also ideally be able to register a callback on accepting completions as well.

One thing we're still ironing out as we implement GitHub Copilot completions is what Tab should actually do. In particular, I _think_ we want to continue allowing users to request "regular" R session completions, and that those should probably be produced either by Tab, or Ctrl + Shift, or both. It might be confusing to have _both_ ghost text + completion popups displaying at the same time, but I think it's reasonable to want access to both in a single R session.

**Comment by justjhong** [comment]
Agree with your points. (1) is ideal especially if it is the only way to get a onDocumentChange listener. The most natural behavior would be asking for a completion on a debounced document change event. The only issue with (1) is that it does require rstudioapi to be more opinionated on how completions should work, and the brittleness of it could be an obstacle in the future.

(2) would work for a MVP of the extension, but as you said it would result in some additional latency since Codeium would just have to pull document contents regularly to check for changes. This seems like the main issue though, so potentially (2) coupled with a `registerOnDocumentChangeHandler` function in the rstudioapi would be sufficient as well.

As for you last comment, I think the way it is handled in VSCode is decent. Basically they have both forms of completion, and the priority is set to the IDE-completions. When the dropdown is cleared w/ Esc, then ghost-text suggestions are offered.

**Comment by justjhong** [comment]
@kevinushey Is there an update as to how the rstudio team will proceed with this issue? Thanks again for considering this for your roadmap!

**Comment by kevinushey** [comment]
We have this work targeted for the upcoming RStudio release, but we haven't made a decision yet. I'm thinking we'll start by implementing (2) because of time constraints, but make sure we leave the door open for (1) if we find latency to be an issue.

**Comment by justjhong** [comment]
Thanks for the quick reply. Starting with (2) sounds good to me. The `getDocumentContext` function should be enough for the addin to keep track of state and request completions accordingly. Once there's a PR it would be great to be looped in to make sure the ghost text function has sufficient functionality to make things work!

**Comment by justjhong** [comment]
@kevinushey Hey Kevin, would you mind giving an update on this issue? Also, in the latest release, is there anything new in the rstudioapi that would benefit support for alternative autocomplete?

**Comment by kevinushey** [comment]
@justjhong I've put together a candidate PR in https://github.com/rstudio/rstudio/pull/13616. This basically adds two things:

- An API function `.rs.api.setGhostText()`, for setting ghost text at the cursor position to some value;
- A set of R options that can be set, to allow users to be informed when documents open in RStudio have changed.

Would you be able to take a look at the PR and let me know if this would be sufficient for a first start in your case?

**Comment by justjhong** [comment]
Hi @kevinushey, thank you very much for the quick turnaround! I took a look at the PR and it generally looks good for our purpose, but I have a couple of questions/comments:

1. It looks like `setGhostText` only takes text as an argument. Does this imply that the inserted ghost text will always be where the cursor is? In some cases, we would like to interleave ghost text, which may require finer control of the ghost text placement (see below). Would this be possible by adding an optional range argument?
<img width="405" alt="image" src="https://github.com/rstudio/rstudio/assets/14086852/40553790-fada-4558-ba7e-e4e294d52302">
2. What is the behavior of the ghost text exactly? Does it disappear on clicking/moving away from the original position (this is the preferred behavior)? Does it properly avoid being committed to the file on save?
3. I don't have a strong opinion on options vs. getHook, so whatever is easiest and functions as you described should be fine. I'm assuming that I can call something like `getSourceEditorContext` to get the full context within the callback.

Thanks again Kevin! This is super helpful.

**Comment by kevinushey** [comment]
> It looks like setGhostText only takes text as an argument. Does this imply that the inserted ghost text will always be where the cursor is? In some cases, we would like to interleave ghost text, which may require finer control of the ghost text placement (see below). Would this be possible by adding an optional range argument?

Unfortunately, RStudio's editor doesn't currently support interleaving of ghost text with the original document contents in the same way as e.g. VSCode, so this is part of the reason why. We might still want to allow the API to accept a document range that the ghost text would insert / replace, though...

> What is the behavior of the ghost text exactly? Does it disappear on clicking/moving away from the original position (this is the preferred behavior)? Does it properly avoid being committed to the file on save?

Yep, all of those are true. The ghost text never actually exists in the document contents / buffer; it's purely a visual UI display.

**Comment by justjhong** [comment]
@kevinushey Thanks for the reply!

> Unfortunately, RStudio's editor doesn't currently support interleaving of ghost text with the original document contents in the same way as e.g. VSCode, so this is part of the reason why. We might still want to allow the API to accept a document range that the ghost text would insert / replace, though...
Got it, so I'm guessing there is a single variable that controls the ghost text existence and placement such that multiple calls to setGhostText would override each other?

> Yep, all of those are true. The ghost text never actually exists in the document contents / buffer; it's purely a visual UI display.
That's great. Does this work for multiline suggestions?

**Comment by kevinushey** [comment]
> Got it, so I'm guessing there is a single variable that controls the ghost text existence and placement such that multiple calls to setGhostText would override each other?

Exactly -- there can only be one instance of ghost text in the editor at a time, and so calling it multiple times would replace the old ghost text.

> That's great. Does this work for multiline suggestions?

It does. For reference, the display as follows:

<img width="552" alt="Screenshot 2023-09-11 at 12 08 38 PM" src="https://github.com/rstudio/rstudio/assets/1976582/1724e609-4e82-4af1-8264-ffde926254cc">

**Comment by justjhong** [comment]
Got it, in that case I think it's safe to move forward with this PR! I can just omit interleaved completions entirely for now. Once it is merged, is there a way I can run the commit with the change before it get's officially released?

**Comment by jgutman** [comment]
@justjhong yes, once it is merged, it will be available in the RStudio daily build, which you can download and install here: https://dailies.rstudio.com/ We are still deciding whether it will be in the [Desert Sunflower daily build](https://dailies.rstudio.com/rstudio/desert-sunflower/), or the subsequent one which will show up shortly -- can confirm the link once it's merged

**Comment by justjhong** [comment]
@kevinushey @jgutman Thank you for moving this issue along! I saw the PR was merged into the desert sunflower branch. For me to call the `setGhostText` function, is there a way I can call the function from an extension at the moment, or would there need to be an associated `rstudioapi` PR merged first?

**Comment by jgutman** [comment]
For now you should be able to call with .rs.api.setGhostText function when using a Desert Sunflower daily build. Eventually we can add it to the rstudioapi package but that hasn't been done yet

**Comment by justjhong** [comment]
Fantastic, thank you @jgutman!

**Comment by ronblum** [comment]
@justjhong This should be included in the new release, [Desert Sunflower 2023.09.0+463](https://posit.co/download/rstudio-desktop/), if you'd like to try it out. Let us know how it goes!

**Comment by justjhong** [comment]
Thanks @ronblum! I already tried it with the daily release, and it worked great! I was able to get a working demo up easily. The only trouble I faced was sometimes the normal autocomplete would pop up at the same time, which was some behavior I saw on another thread, I believe. But its a relatively minor issue.

**Comment by ronblum** [comment]
@justjhong In case you still want to keep both Copilot suggestions and autocompletion, <kbd>Tab</kbd> will accept the Copilot suggestion and <kbd>Ctrl-Space</kbd> will accept the autocompletion.

**Comment by justjhong** [comment]
@kevinushey @ronblum thanks for clarifying! These are good choices to modify the behavior. Ill make it clear to users and leave it to them.

**Comment by justjhong** [comment]
Yes! Feel free to close, this should be enough to support Codeium, and it's quite performant! If anything comes up do you prefer I reopen this or create a new issue?

**Comment by ronblum** [comment]
Open a new issue, since this one is specifically about the implementation. We usually try to keep one request—bug or enhancement—per issue.

Thanks for confirming, and I hope you like it!

**Comment by jthomasmock** [comment]
Hi @justjhong - did you have a repo where we could try out the Codeium add-in?

**Comment by justjhong** [comment]
Hi @jthomasmock, I do! However, it is still private since its WIP (to avoid confusion). Though, I would be happy to add you as a collaborator if you would take a look.

**Comment by jthomasmock** [comment]
> Hi @jthomasmock, I do! However, it is still private since its WIP (to avoid confusion). Though, I would be happy to add you as a collaborator if you would like take a look.

That would be great! Happy to help kick the tires 🤝

**Comment by maelle** [comment]
@justjhong I'm curious too! :slightly_smiling_face:

**Comment by mgazzin** [comment]
Hello, I am interested too! Is this change going to production release?

**Comment by ShellyShen1019** [comment]
The idea that Support for Codeium autocomplete as a Github Copilot alternative would be great! Happy to help kick the tire

**Comment by bbolker** [comment]
any updates on this?  I can't find anything else about it (but maybe I'm not looking in the right place ...)

**Comment by justjhong** [comment]
The support for ghosttext has been added by the Rstudio team, which makes the addin feasible, but the final version is indefinitely stalled. It mostly works just without a pretty frontend. I'll add the folks who commented above as collaborators to the private repo in the meantime.

**Comment by maelle** [comment]
thank you!

**Comment by jrausch12** [comment]
@justjhong any updates on this or another thread I can follow?
